cdap-app-templates/cdap-etl/cdap-etl-core/src/main/java/co/cask/cdap/etl/planner/ControlDag.java

```diff
@@ -0,0 +1,301 @@
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Multiset;
+
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.TreeSet;
+import java.util.UUID;
+
+/**
+ * A DAG (directed acyclic graph) where edges represent a happens-before relationship.
+ * In these types of scenarios, certain edges may be redundant and can be removed.
+ * This simplifies the dag into something that is much easier to work with if it needs
+ * to be used as a fork-join type of dag for workflow execution.
+ */
+public class ControlDag extends Dag {
+  private static final Set<String> EMPTY = ImmutableSet.of();
+  private final Multiset<String> nodeVisits;
+
+  public ControlDag(Collection<Connection> connections) {
+    super(connections);
+    this.nodeVisits = HashMultiset.create();
+  }
+
+  /**
+   * Record that this node was visited.
+   *
+   * @param node node that was visited
+   * @return the number of times this node was visited, including the visit from this call
+   */
+  public int visit(String node) {
+    nodeVisits.add(node);
+    return nodeVisits.count(node);
+  }
+
+  /**
+   * Resets the number of times each node was visited back to 0.
+   */
+  public void resetVisitCounts() {
+    nodeVisits.clear();
+  }
+
+  /**
+   * Flattens the control dag to remove connections between branches of different forks, which would
+   * make the dag unusable in pure fork-join workflows.
+   *
+   * For example the following dag is not a fork-join dag:
+   *
+   *      |--> n2 -------|
+   *      |              |--> n5
+   *      |--> n3 -------|
+   *  n1--|           |
+   *      |           v
+   *      |--> n4 --> n6
+   *
+   *  There are many ways to turn this a fork-join while still respecting all happens-before relationships,
+   *  but for simplicity we'll use an algorithm that doesn't have any nested forks and will turn the above into:
+   *
+   *       |--> n2 --|
+   *       |         |               |--> n5 --|
+   *  n1 --|--> n3 --|--> n2.n3.n4 --|         |--> n5.n6
+   *       |         |               |--> n6 --|
+   *       |--> n4 --|
+   *
+   *  The algorithm is to insert a join node whenever it sees a fork. Every time there is a fork, we will follow
+   *  each branch to its endpoint (a node that forks, merges, or is a sink), then insert a join node that each
+   *  branch endpoint connects to.
+   */
+  public void flatten() {
+    trim();
+    String source;
+    // if we have multiple sources, insert a fork node as the new source
+    if (sources.size() > 1) {
+      // copy to avoid concurrent modification
+      Set<String> sourcesCopy = new HashSet<>(sources);
+      String newId = generateJoinNodeName(sourcesCopy);
+      addNode(newId, EMPTY, sourcesCopy);
+      source = newId;
+    } else {
+      source = sources.iterator().next();
```

Comment:
if there aren't any sources, it means there is a cycle somewhere, which should have been checked before this point. Will add a comment.